cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/RemoteTaskExecutor.java

```diff
@@ -80,35 +90,87 @@ public RemoteTaskExecutor(CConfiguration cConf, RemoteClientFactory remoteClient
+            throw RemoteExecutionException.fromBasicThrowable(basicThrowable);
+          }
+          return new RetryableTaskResult(getResponseBody(httpResponse), attempt);
+        } catch (NoRouteToHostException e) {
           throw new RetryableException(
-            String.format("Received response code %s for %s", httpResponse.getResponseCode(),
-                          runnableTaskRequest.getClassName()));
-        }
-        if (httpResponse.getResponseCode() != HttpURLConnection.HTTP_OK) {
-          BasicThrowable basicThrowable = GSON
-            .fromJson(new String(getResponseBody(httpResponse)), BasicThrowable.class);
-          throw RemoteExecutionException.fromBasicThrowable(basicThrowable);
+            String.format("Received exception %s for %s", e.getMessage(), runnableTaskRequest.getClassName()));
         }
-        return getResponseBody(httpResponse);
-      } catch (NoRouteToHostException e) {
-        throw new RetryableException(
-          String.format("Received exception %s for %s", e.getMessage(), runnableTaskRequest.getClassName()));
+      }, retryStrategy, Retries.DEFAULT_PREDICATE);
+      //emit metrics with successful result
+      emitMetrics(startTime, true, runnableTaskRequest, retryableTaskResult.getAttempts());
+      return retryableTaskResult.getResult();
+    } catch (Exception e) {
+      //emit metrics with failed result
+      emitMetrics(startTime, false, runnableTaskRequest, getAttempts(e));
+      throw e;
+    }
+  }
+
+  /**
+   * Find if attempt count is included in the suppressed throwable
+   *
+   * @param e Exception to analyze
+   * @return attempt count
+   */
+  private int getAttempts(Exception e) {
+    int attempts = 0;
+    Throwable[] suppressed = e.getSuppressed();
+    for (Throwable t : suppressed) {
+      if (t instanceof RetryFailedException) {
+        attempts = ((RetryFailedException) t).getRetries();
+        break;
+      } else if (t instanceof RetriesExhaustedException) {
+        attempts = ((RetriesExhaustedException) t).getRetries();
+        break;
       }
-    }, retryStrategy);
+    }
+    return attempts;
+  }
+
+  private void emitMetrics(long startTime, boolean success, RunnableTaskRequest runnableTaskRequest, int attempts) {
+    String taskClass = getTaskClassName(runnableTaskRequest);
+    Map<String, String> metricTags = new HashMap<>();
+    metricTags.put(Constants.Metrics.Tag.CLASS, taskClass);
+    metricTags.put(Constants.Metrics.Tag.STATUS, success ? "success" : "failure");
+    metricTags.put(Constants.Metrics.Tag.TRIES, String.valueOf(attempts));
+    metricsCollectionService.getContext(metricTags).increment(Constants.Metrics.TaskWorker.CLIENT_REQUEST_COUNT, 1L);
+    metricsCollectionService.getContext(metricTags)
+      .gauge(Constants.Metrics.TaskWorker.CLIENT_REQUEST_LATENCY_MS, System.currentTimeMillis() - startTime);
+    LOG.debug("Emitting metric tags {} with service {}", metricTags, metricsCollectionService);
```

Comment:
This seems more a trace log. Is it needed?

Reply:
No, was using it for debugging. Removed.